### ISSUE TITLE
Card stuck in builder

### ISSUE THREAD

**Issue by pixelzoom** [phetsims/function-builder#50]
Moved this issue here from https://github.com/phetsims/function-builder/issues/39#issuecomment-220080648.

@phet-steele reported:
> @pixelzoom the second procedure is indeed fixed, but now I can get cards stuck in the function machine. NO idea how, will have to look into it. This is FF.
![stuckcard](https://cloud.githubusercontent.com/assets/5863899/15366607/5a5de116-1ce2-11e6-826f-13dfe64710d0.PNG)

**Comment by pixelzoom** [comment]
Additional relevant comments from #39:

@phet-steele:
> @pixelzoom I'm honestly at a loss. My only guess is it's tied to a split second performance drop in FF that happens soon after sim start up (can't do this in Chrome). If I'm lucky enough to release the card while simultaneously trying to invert it, and the stars align causing a performance drop at that precise moment, then maybe, just maybe, the card will get stuck. The card can't actually be inverted though, so I guess the main issue here is resolved.

@phet-steele:
> I grabbed a screenshot of fps over time (recorded in the profiler). There is a quick yet noticeable drop soon after the sim finishes loading.
![funperf](https://cloud.githubusercontent.com/assets/5863899/15404819/89cc3580-1dbc-11e6-88ca-2e4456682f09.PNG)

@pixelzoom:
> Are you referring to the drop at ~15 seconds?  Does this drop occur if the sim is simply idle, or while you are doing something?

@phet-steele:
> Yes at 15, and I think I have to at least enter a screen for the drop. Profiler shows no drop while idle on the home screen. Also should mention this is current master.

**Comment by pixelzoom** [comment]
I've been trying to reproduce this with FF 46.01 on Mac OS 10.11.5, and I've been unsuccessful.  

@phet-steel:
Is there anything else you can tell me about the test machine? FF version? Mac or Win? Touch screen?  

Any other details about what you're doing when this happens? How long after startup? How many attempts at dragging card through? Video?...

**Comment by pixelzoom** [comment]
Skype:

[5/23/16, 10:49:02 AM] Steele Dalton: you say you cant make it happen on OSX 10.11.5, but are you at least seeing a performance drop? I find it easiest to see by quickly grabbing a card and moving it back and forth. Eventually the card will lag behind my mouse for a split second
[5/23/16, 10:55:00 AM] Chris Malley: I see no noticeable performance drop. Are you testing on Mac?
[5/23/16, 10:56:49 AM] Steele Dalton: the reported comments are from Win/Chrome, BUT i do see it on Mac

**Comment by phet-steele** [comment]
FF version 46.0.1
Win 10 & Mac OSX 10.11.5
No touch screen

Hard to say how long after startup. It isn't always the same amount of time, but I've never seen it happen later than 10 seconds after loading. It only takes the one lucky drag to get the card stuck, as long as the release happens at the performance drop. I try to release while dragging the card left near the center of the function machine. @pixelzoom I'm inclined to say this should be a lower priority, it's very hard to duplicate and performance stabilizes quickly.

**Comment by pixelzoom** [comment]
One more question @phet-steele ...  After the card gets stuck, does toggling the 'See Inside' window on and off cause the card to move to the output carousel?

**Comment by phet-steele** [comment]
After 36 minutes of trying, I made it happen:

https://drive.google.com/file/d/0B3HJopSo_QqLVVdXYzBhZld6UXc/view?usp=sharing

> After the card gets stuck, does toggling the 'See Inside' check box on and off cause the card to move to the output carousel? Or does it remain stuck?

Short answer....it moves! Happens near 18 secs in the video.

**Comment by pixelzoom** [comment]
I still can't reproduce this.  @phet-steele do you recall if there was anything in the console?

**Comment by phet-steele** [comment]
@pixelzoom I do recall, and there was nothing.

**Comment by pixelzoom** [comment]
The card is definitely positioned where a 'see inside' window is.  And toggling the 'see inside' check box cause the card to flush out of the carousel.   So while, I can't reproduced this, I can look in CardNode's drag handler for a situation where a card is moved to a 'see inside' window without consulting the state of `seeInsideProperty`.

**Comment by pixelzoom** [comment]
Skype chat with @phet-steele:

[6/10/16, 10:54:52 AM] Chris Malley: In your both your screenshot and video for the issue, you have a non-invertible function in the builder.  Did you ever reproduce this when a non-invertible function was not present?  Significant because it's a different code path for non-invertible functions.
[6/10/16, 10:58:40 AM] Steele Dalton: good question, and i think i only ever used the camera. I think my procedure would be very difficult if i had an invertible function in place.
[6/10/16, 10:58:46 AM] Chris Malley: Did the card get stuck while you were still dragging, or after you released the pointer?
[6/10/16, 10:58:56 AM] Chris Malley: hard to tell from the video.
[6/10/16, 10:59:29 AM] Steele Dalton: release
[6/10/16, 10:59:41 AM] Chris Malley: great. every drag handler has separate chunks of code for "startDrag", "dragging" and "endDrag"
[6/10/16, 10:59:50 AM] Chris Malley: so that narrows it down for where to look.

So I should inspect `options.endDrag`, which starts at `CardNode` line 181.

**Comment by pixelzoom** [comment]
More Skype chat:

> [6/10/16, 11:05:04 AM] Chris Malley: it looks like the card gets stuck in these conditions: it's blocked by a non-invertible function, it's released, it starts moving to the right, and it stops at a 'see inside' window that's not there. I'm basing this on what I see in the video.  Does it sound like a reasonable hypothesis?
[6/10/16, 11:06:53 AM] Steele Dalton: yea

**Comment by pixelzoom** [comment]
Consider adding some debug output that is placed based on "educated guess" about what might be happening, then have @phet-steele reproduce and capture console output.

**Comment by phet-steele** [comment]
@andrea-phet just confirmed this is not FF specific. She duplicated on IE11 and Chrome (on Win 7)

**Comment by pixelzoom** [comment]
@phet-steele @andrea-phet When it was duplicated on IE11 and Chrome (Win7), did it always involve a non-intervertible function in the builder?

**Comment by andrea-phet** [comment]
It was only tested with a non-invertible function. It would be hard to make it get stuck with an invertible function because the card would just flow to the other side. And I noticed I made the card get stuck after "bouncing" the card against the function, "catching" it, and repeat.

**Comment by pixelzoom** [comment]
Thanks @andrea-phet.

**Comment by pixelzoom** [comment]
@phet-steele It's highly likely that the fix for https://github.com/phetsims/function-builder/issues/40#issuecomment-225695603 applies to this issue.  But since I still haven't been able to reproduce this issue, I'm not certain.  Could you please test this dev version to see if the issue is still present?
http://www.colorado.edu/physics/phet/dev/html/function-builder/1.0.0-dev.54/function-builder_en.html

**Comment by phet-steele** [comment]
@pixelzoom the group spent a combined hour on this issue, with no luck in reproducing. Since the only way to verify this issue is to spend a vast amount of time trying to reproduce it, I think an hour is good enough. So my verdict is it's fixed.

**Comment by pixelzoom** [comment]
Thanks group! Closing.
